analyzers/src/SonarAnalyzer.CSharp/Rules/UnchangedLocalVariablesShouldBeConst.cs

```diff
@@ -154,16 +160,35 @@ private static bool HasMutableUsagesInMethod(VariableDeclaratorSyntax variable,
                 Equals(variableSymbol, semanticModel.GetSymbolInfo(id).Symbol);
         }
 
-        private static bool IsMutatingUse(IdentifierNameSyntax identifier) =>
+        private static bool IsMutatingUse(SemanticModel semanticModel, IdentifierNameSyntax identifier) =>
             identifier.Parent switch
             {
                 AssignmentExpressionSyntax assignmentExpression => Equals(identifier, assignmentExpression?.Left),
                 ArgumentSyntax argumentSyntax => argumentSyntax.IsInTupleAssignmentTarget() || !argumentSyntax.RefOrOutKeyword.IsKind(SyntaxKind.None),
                 PostfixUnaryExpressionSyntax _ => true,
                 PrefixUnaryExpressionSyntax _ => true,
-                _ => false
+                _ => IsUsedAsLambdaExpression(semanticModel, identifier)
             };
 
+        private static bool IsUsedAsLambdaExpression(SemanticModel semanticModel, IdentifierNameSyntax identifier)
+        {
+            if (identifier.FirstAncestorOrSelf<LambdaExpressionSyntax>().GetSelfOrTopParenthesizedExpression() is { } lambda)
+            {
+                if (lambda.Parent is ArgumentSyntax argument
+                    && argument.FirstAncestorOrSelf<InvocationExpressionSyntax>() is { } invocation)
+                {
+                    var lookup = new CSharpMethodParameterLookup(invocation, semanticModel);
+                    return lookup.TryGetSymbol(argument, out var parameter) && parameter.IsType(KnownType.System_Linq_Expressions_Expression_T);
+                }
+                else if (lambda.Parent is AssignmentExpressionSyntax assignment // Lambda cannot be on the left side
+                    && semanticModel.GetSymbolInfo(assignment.Left).Symbol is { } assignmentTargetSymbol)
+                {
+                    return assignmentTargetSymbol.GetSymbolType().Is(KnownType.System_Linq_Expressions_Expression_T);
+                }
```

Comment:
We can use `assignment.Left.IsKnownType(KnownType.System_Linq_Expressions_Expression_T, semanticModel)` and simplify the code a bit.

Reply:
I don't agree on the `a bit`. It's `quite a lot` 😄